runtime/compiler/z/codegen/J9CodeGenerator.cpp

```diff
@@ -3837,11 +3839,54 @@ J9::Z::CodeGenerator::inlineDirectCall(
 
          if ((!TR::Compiler->om.canGenerateArraylets() || node->isUnsafeGetPutCASCallOnNonArray()) && node->isSafeForCGToFastPathUnsafeCall())
             {
-            resultReg = TR::TreeEvaluator::VMinlineCompareAndSwap(node, cg, (comp->useCompressedPointers() ? TR::InstOpCode::CS : TR::InstOpCode::getCmpAndSwapOpCode()), IS_OBJ);
+            resultReg = TR::TreeEvaluator::VMinlineCompareAndSwap(node, cg, (comp->useCompressedPointers() ? TR::InstOpCode::CS : TR::InstOpCode::getCmpAndSwapOpCode()), IS_OBJ, false);
             return true;
             }
          break;
 
+      case TR::jdk_internal_misc_Unsafe_compareAndExchangeInt:
+         if ((!TR::Compiler->om.canGenerateArraylets() || node->isUnsafeGetPutCASCallOnNonArray()) && node->isSafeForCGToFastPathUnsafeCall())
+            {
+            if (!disableCAEIntrinsic)
+               {
+               resultReg = TR::TreeEvaluator::VMinlineCompareAndSwap(node, cg, TR::InstOpCode::CS, IS_NOT_OBJ, true);
+               return true;
+               }
+            }
+         break;
+
+      case TR::jdk_internal_misc_Unsafe_compareAndExchangeLong:
+         if (comp->target().is64Bit() && (!TR::Compiler->om.canGenerateArraylets() || node->isUnsafeGetPutCASCallOnNonArray()) && node->isSafeForCGToFastPathUnsafeCall())
+            {
+            if (!disableCAEIntrinsic)
+               {
+               resultReg = TR::TreeEvaluator::VMinlineCompareAndSwap(node, cg, TR::InstOpCode::CSG, IS_NOT_OBJ, true);
+               return true;
+               }
+            }
+         // Too risky to do Long-31bit version now.
```

Comment:
We should move this comment close to where we only do the transformation for 64-bit.

Reply:
I'm not exactly sure when you want me to move this. The comment is right after the block with the `comp->target().is64Bit()` check.

Reply:
I think right before the condition for 64-bit would be the appropriate place. I would not have thought of `else` part for 31-bit.

Reply:
I find comment would be more relevant around/before the condition `comp->target().is64Bit()`